[Viewport Clipping] [Part 2/4] Add a common base class for ScrollingTree{Sticky|Fixed}Node

whsieh opened the pull request:
#### 57a561ae10fa118c6f17ecfdf593be67f986061e
<pre>
[Viewport Clipping] [Part 2/4] Add a common base class for ScrollingTree{Sticky|Fixed}Node
<a href="https://bugs.webkit.org/show_bug.cgi?id=292109">https://bugs.webkit.org/show_bug.cgi?id=292109</a>
<a href="https://rdar.apple.com/150116251">rdar://150116251</a>

Reviewed by Abrar Rahman Protyasha.

In preparation for a configuration where viewport-constrained objects in the mainframe are clipped
to the layout viewport rect, we introduce a base class — `ScrollingTreeViewportConstrainedNode` —
from which `ScrollingTreeStickyNode` and `ScrollingTreeFixedNode` both derive. See below for more
details.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Add the new files.

* Source/WebCore/page/scrolling/ScrollingTreeFixedNode.cpp:
(WebCore::ScrollingTreeFixedNode::ScrollingTreeFixedNode):
(WebCore::ScrollingTreeFixedNode::computeLayerPosition const): Deleted.

Move this method into `ScrollingTreeViewportConstrainedNode` (see below).

* Source/WebCore/page/scrolling/ScrollingTreeFixedNode.h:
* Source/WebCore/page/scrolling/ScrollingTreeStickyNode.cpp:
(WebCore::ScrollingTreeStickyNode::ScrollingTreeStickyNode):
(WebCore::ScrollingTreeStickyNode::computeAnchorLayerPosition const):

Rename the existing method `computeLayerPosition` to `computeAnchorLayerPosition`. This seems
unnecessary at the moment, but will become important once the clipping layer is introduced, since
the position of the clipping layer needs to be computed separately from the position of the anchor
layer.

Namely, for a viewport-clipped sticky node that does not yet intersect with the constraining rect
(and therefore behaves as relative instead of fixed), we need to blit the clipping layer so it
matches the viewport rect while scrolling, and simultaneously counter-blit the anchor layer
underneath such that the sticky contents (visually) scroll with the rest of the page. Once the
constraining rect is reached and the sticky node behaves like a fixed-position node, only the
clipping layer continues blitting, while the anchor node offset becomes constant.

(WebCore::ScrollingTreeStickyNode::scrollDeltaSinceLastCommit const):
(WebCore::ScrollingTreeStickyNode::computeLayerPosition const): Deleted.
* Source/WebCore/page/scrolling/ScrollingTreeStickyNode.h:
* Source/WebCore/page/scrolling/ScrollingTreeViewportConstrainedNode.cpp: Copied from Source/WebCore/page/scrolling/ScrollingTreeFixedNode.cpp.
(WebCore::ScrollingTreeViewportConstrainedNode::ScrollingTreeViewportConstrainedNode):

Consolidate logic to call `fixedOrStickyNodeAdded` when creating a viewport-constrained node.

(WebCore::ScrollingTreeViewportConstrainedNode::computeLayerPosition const):

Move this logic into `ScrollingTreeViewportConstrainedNode`, where it can be used for both sticky
and fixed nodes.

* Source/WebCore/page/scrolling/ScrollingTreeViewportConstrainedNode.h: Copied from Source/WebCore/page/scrolling/ScrollingTreeFixedNode.h.
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeFixedNodeCocoa.h:
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeStickyNodeCocoa.h:
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeStickyNodeCocoa.mm:
(WebCore::ScrollingTreeStickyNodeCocoa::applyLayerPositions):
* Source/WebCore/page/scrolling/coordinated/ScrollingTreeStickyNodeCoordinated.cpp:
(WebCore::ScrollingTreeStickyNodeCoordinated::applyLayerPositions):

Canonical link: <a href="https://commits.webkit.org/294285@main">https://commits.webkit.org/294285@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/603ae2bd7d1b31cb4ed9749a8f6b19bf9ec2bb3a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101334 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20996 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11300 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106485 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51961 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103375 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21305 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29491 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77177 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34215 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104341 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16433 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91524 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57524 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16252 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9536 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51309 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86142 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9610 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108838 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28462 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20943 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86152 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28824 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87731 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85712 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21807 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30440 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8145 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22573 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28392 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33666 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28204 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31524 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29762 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/scrolling/ScrollingTreeViewportConstrainedNode.h:
```diff
@@ -0,0 +1,51 @@
+    ScrollingTreeViewportConstrainedNode(ScrollingTree&, ScrollingNodeType, ScrollingNodeID);
+
+    virtual ~ScrollingTreeViewportConstrainedNode();
+    virtual const ViewportConstraints& constraints() const = 0;
```
Nit: remove the `ScrollingConstraints.h` include and forward declare `ViewportConstraints`.

whsieh replied:
Fixed — thanks for the review!

whsieh committed:
[Viewport Clipping] [Part 2/4] Add a common base class for ScrollingTree{Sticky|Fixed}Node
https://bugs.webkit.org/show_bug.cgi?id=292109
rdar://150116251

Reviewed by Abrar Rahman Protyasha.

In preparation for a configuration where viewport-constrained objects in the mainframe are clipped
to the layout viewport rect, we introduce a base class — `ScrollingTreeViewportConstrainedNode` —
from which `ScrollingTreeStickyNode` and `ScrollingTreeFixedNode` both derive. See below for more
details.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Add the new files.

* Source/WebCore/page/scrolling/ScrollingTreeFixedNode.cpp:
(WebCore::ScrollingTreeFixedNode::ScrollingTreeFixedNode):
(WebCore::ScrollingTreeFixedNode::computeLayerPosition const): Deleted.

Move this method into `ScrollingTreeViewportConstrainedNode` (see below).

* Source/WebCore/page/scrolling/ScrollingTreeFixedNode.h:
* Source/WebCore/page/scrolling/ScrollingTreeStickyNode.cpp:
(WebCore::ScrollingTreeStickyNode::ScrollingTreeStickyNode):
(WebCore::ScrollingTreeStickyNode::computeAnchorLayerPosition const):

Rename the existing method `computeLayerPosition` to `computeAnchorLayerPosition`. This seems
unnecessary at the moment, but will become important once the clipping layer is introduced, since
the position of the clipping layer needs to be computed separately from the position of the anchor
layer.

Namely, for a viewport-clipped sticky node that does not yet intersect with the constraining rect
(and therefore behaves as relative instead of fixed), we need to blit the clipping layer so it
matches the viewport rect while scrolling, and simultaneously counter-blit the anchor layer
underneath such that the sticky contents (visually) scroll with the rest of the page. Once the
constraining rect is reached and the sticky node behaves like a fixed-position node, only the
clipping layer continues blitting, while the anchor node offset becomes constant.

(WebCore::ScrollingTreeStickyNode::scrollDeltaSinceLastCommit const):
(WebCore::ScrollingTreeStickyNode::computeLayerPosition const): Deleted.
* Source/WebCore/page/scrolling/ScrollingTreeStickyNode.h:
* Source/WebCore/page/scrolling/ScrollingTreeViewportConstrainedNode.cpp: Copied from Source/WebCore/page/scrolling/ScrollingTreeFixedNode.cpp.
(WebCore::ScrollingTreeViewportConstrainedNode::ScrollingTreeViewportConstrainedNode):

Consolidate logic to call `fixedOrStickyNodeAdded` when creating a viewport-constrained node.

(WebCore::ScrollingTreeViewportConstrainedNode::computeLayerPosition const):

Move this logic into `ScrollingTreeViewportConstrainedNode`, where it can be used for both sticky
and fixed nodes.

* Source/WebCore/page/scrolling/ScrollingTreeViewportConstrainedNode.h: Copied from Source/WebCore/page/scrolling/ScrollingTreeFixedNode.h.
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeFixedNodeCocoa.h:
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeStickyNodeCocoa.h:
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeStickyNodeCocoa.mm:
(WebCore::ScrollingTreeStickyNodeCocoa::applyLayerPositions):
* Source/WebCore/page/scrolling/coordinated/ScrollingTreeStickyNodeCoordinated.cpp:
(WebCore::ScrollingTreeStickyNodeCoordinated::applyLayerPositions):

Canonical link: https://commits.webkit.org/294285@main
